RxSwift/Traits/Single.swift

```diff
@@ -248,4 +248,35 @@ extension PrimitiveSequenceType where TraitType == SingleTrait {
         -> Single<R> {
             return Single<R>(raw: primitiveSequence.source.flatMap(selector))
     }
+    
+    /**
+     Merges the specified observable sequences into one observable sequence by using the selector function whenever all of the observable sequences have produced an element at a corresponding index.
+     
+     - parameter resultSelector: Function to invoke for each series of elements at corresponding indexes in the sources.
+     - returns: An observable sequence containing the result of combining elements of the sources using the specified result selector function.
+     */
+    public static func zip<C: Collection, R>(_ collection: C, _ resultSelector: @escaping ([ElementType]) throws -> R) -> PrimitiveSequence<TraitType, R> where C.Iterator.Element == PrimitiveSequence<TraitType, ElementType> {
+        
+        if collection.isEmpty {
+            return PrimitiveSequence<TraitType, R>(raw: .just(try! resultSelector([])))
```

Comment:
We can't use `try!`.

We should probably do something as this.

```
return PrimitiveSequence<TraitType, R>.defer {
     return PrimitiveSequence<TraitType, R>(raw: .just(try resultSelector([])))
}
```

and add a test case for that.

Defer is here to make sure `resultSelector ` is being called once per subscription and not per zip invocation. This shouldn't have much difference, but if somebody decides to do a dirty hack and perform side effects inside `resultSelector`, version with `defer` will work with `retry` operators nicely.

Reply:
gotcha! will fix it

Reply:
just one question. how should it be checked if the resultSelector is called not once per invocation but once per subscription? I'm not so used to tests in Rx.